### ISSUE TITLE
Door near the smithy in the castle of the Old Camp is inaccessible

### ISSUE THREAD

**Issue by AmProsius** [AmProsius/gothic-1-community-patch#46]
**Describe the bug**
The smith's door cannot not be opened, because it references a key that does not exist.

**Expected behavior**
The door near the smithy in the castle of the Old Camp is now accessible with the correct key which can now be obtained in Stone's cell in the dungeon of the Old Camp.

**Additional context**
The key to the smithy's door does not work because the door references a wrong key instance name: `ITKEY_OB_SMITH_01` does not exist. It needs to be changed to `ItKe_OB_Smith_01`.

![door](https://user-images.githubusercontent.com/20203034/109395183-587be000-792b-11eb-99bc-54e7a7d045cb.png)

**Implementation**
This issue was merged with #47. We supply our own G1CP copy of the key instance to allow saving and loading. When removing G1CP the then invalid instance will automatically vanish on loading. If reinstalling G1CP the key instance will be re-added (given the conditions to do so are met).

### Apply
```
1. Is door valid (identified,
   locked, has incorrect key instance)
     │             │
     │ True        │ False
     │             └─────────────────────────────────────────────┐
     V                                                           │
2. No such key instance exists                                   │
   (could have been added by a mod to                            │
    fix the issue the other way around)                          │
     │             │                                             │
     │ True        │ False                                       │
     │             └──────────────────────────────────────────┐  │
     V                                                        │  │
3. Does the G1CP key already exist in the world               │  │
     │             │                                          │  │
     │ True        │ False                                    │  │
     │             │                                          │  │
     │             V                                          │  │
     │        A. Can key be inserted?                         │  │
     │           (waypoint exists)                            │  │
     │             │            │                             │  │
     │             │ True       │ False                       │  │
     │             │            └──────────────────────────┐  │  │
     │             V                                       │  │  │
     │        B. Is story progressed enough?               │  │  │
     │           (ExploreSunkenTower == True)              │  │  │
     │             │            │                          │  │  │
     │             │ True       │ False                    │  │  │
     │             │            │                          │  │  │
     │             │            V                          │  │  │
     │             │     I.  Will story variable be set    │  │  │
     │             │         in expected function?         │  │  │
     │             │         (ExploreSunkenTower           │  │  │
     │             │          mentioned in                 │  │  │
     │             │          B_Story_ExploreSunkenTower)  │  │  │
     │             │            │           │              │  │  │
     │             │            │ True      │ False        │  │  │
     │             │            │           └───────────┐  │  │  │
     │             │            V                       │  │  │  │
     │             │     II. Hook after the story       │  │  │  │
     │             │         function to run this here  │  │  │  │
     │             │         function again and to      │  │  │  │
     │             │         insert the key then        │  │  │  │
     │             │            │                       │  │  │  │
     │             │            └────────────────────┐  │  │  │  │
     │             │                                 │  │  │  │  │
     │             │                                 │  │  │  │  │
     │             V                                 │  │  │  │  │
     │        C. Insert the key                      │  │  │  │  │
     │           (in Stone's cell)                   │  │  │  │  │
     │             │           │                     │  │  │  │  │
     │             │ Success   │ Failed (unlikely)   │  │  │  │  │
     │             │           └──────────────────┐  │  │  │  │  │
     │             │                              │  │  │  │  │  │
     V             V                              │  │  │  │  │  │
4. Adjust the key instance                        │  │  │  │  │  │
   for the door                                   │  │  │  │  │  │
        │                                         │  │  │  │  │  │
        │                                         │  │  │  │  │  │
        V                                         V  V  V  V  V  V
    return True                                      return False
    (fix applied)                                 (fix not applied)
```

### Revert
```
1. Is door valid (identified, has
   updated G1CP key instance)
     │
     │ True
     V
2. Adjust the key instance
   for the door

(G1CP key vanishes automatically if loading a game save without G1CP active)
```

In short, the fix (i.e. adjusting the door) will **only** be applied if
1. the door has the incorrect key instance
2. there is no such key instance in the script
3. the story has progressed far enough

Would the story be changed somewhat (the variables renamed) the entire fix would not be applied.

**Comment by catalinstoian** [comment]
MissionItems_1.d

from
```
//SCHMIEDESCHLÜSSEL ERZBARONE
INSTANCE ItKe_OB_Smith_01(C_Item)
{	
	name 					=	"Iron Key";

	mainflag 				=	ITEM_KAT_NONE;
	flags 					=	0;	

	value 					=	0;

	visual 					=	"ItKe_Key_02.3ds";
	material 				=	MAT_METAL;
	description			= name;
	TEXT[0]				= "Opens the door to the smithy's store.";
	////COUNT[0]			= ;
	//TEXT[1]				= "";
	////COUNT[1]			= ;
	//TEXT[2]				= "";
	//COUNT[2]			= ;
	//TEXT[3] 			= "";
	//COUNT[3]			= ;
	//TEXT[4]				= ""; 
	////COUNT[4]			= ;
	TEXT[5]				= NAME_Value;
	COUNT[5]			= value;
};
```
to
```
//SCHMIEDESCHLÜSSEL ERZBARONE
INSTANCE ITKEY_OB_SMITH_01(C_Item)
{	
	name 					=	"Iron Key";

	mainflag 				=	ITEM_KAT_NONE;
	flags 					=	0;	

	value 					=	0;

	visual 					=	"ItKe_Key_02.3ds";
	material 				=	MAT_METAL;
	description			= name;
	TEXT[0]				= "Opens the door to the smithy's store.";
	////COUNT[0]			= ;
	//TEXT[1]				= "";
	////COUNT[1]			= ;
	//TEXT[2]				= "";
	//COUNT[2]			= ;
	//TEXT[3] 			= "";
	//COUNT[3]			= ;
	//TEXT[4]				= ""; 
	////COUNT[4]			= ;
	TEXT[5]				= NAME_Value;
	COUNT[5]			= value;
};
```

**Comment by szapp** [comment]
Where is this item located in the world? Is it lying around somewhere or does an NPC have in their inventory?

**Comment by AmProsius** [comment]
After digging into this I would suggest to close the issue without changing anything. The label doesn't occur as a bug to me as there are many keys that start with `ItKe_`. There's actually only one exception and that is `ItKey_RB_01`.

As far as I know the key doesn't spawn at all (see #47), so the smithy would remain locked for the entire game.

**Comment by szapp** [comment]
I think the name of this instance matters to work with the door it was made for. If the door specifies a different instance name, the key will not work.

What was the initial purpose of renaming the key? Does it not work on the door it is meant for? In that case, I would suggest to adjust the door-property.

**Comment by AmProsius** [comment]
I can't find any match for `ITKEY_OB_SMITH_01` in the scripts. How can I check the property of the door?

**Comment by szapp** [comment]
@AmProsius That could be done with the spacer. That is quite tedious, however. I can have a look soon, by writing a script and checking in the running game. Could you tell me which is the door in question (a screenshot, waypoint or a rough description - any of those options should suffice)?

**Comment by AmProsius** [comment]
This is the door if I'm not mistaken. The Waypoint is `OCC_STABLE_STAIRCASE_TOP`.

![ScreenShot_2021_2_11_22_9_23](https://user-images.githubusercontent.com/2817152/107699133-e910ba00-6cb5-11eb-9a24-e54d90cc0d0f.jpg)

**Comment by szapp** [comment]
This will be very similar to #52 with respect to adding and removing a certain property of a MOB in the world.

**Comment by szapp** [comment]
In the referenced PR, I decided to lock the door again when reverting the fix (but unlocking it again if it was unlocked before). This should be quite reliable. For details, see #167.

**Comment by szapp** [comment]
This issue will now address both (the original #46) and (the now closed) #47 to *both* adjust the door and insert the key.

**Comment by Quintus24** [comment]
Hi, guys!

Again, I am sorry for spamming a closed issue.

I'm in **Chapter 5** at the moment and I found the **Iron Key** in **Stone's** prison cell.
However, I still can't open the **Smithy's Storage** door.

My character simply stands still when I press and hold the "left mouse button".
The only thing I can do is move the camera up and down while performing this action.

What I observed so far:
 - The character sees the door, as it becomes "**focused**" and the "**Door**" string is displayed.
 - If the character is at a distance from the door but close enough to focus on it, he will shrug his shoulders.
 - Lockpicking the door still works, but the usual strings are displayed (_all over again_ / _picklock has broken_) since the key is mandatory.
 
I tried the "**fix status 46**" command in the console and it said "**Fix # 46 is inactive**".

Is it possible that the fix is not retroactive?

Best Regards,
Quintus24

**Comment by MarkusRamikin** [comment]
To be sure, do I understand correctly that the iron key that I find at Stone's feet in the dungeons is meant to open the door in the smithy, on the right side? Or did you guys give up on that?

If so, I also report that this doesn't work. In this savegame I have freed Stone and I have the key, but I cannot open that door: [savegame3.zip](https://github.com/AmProsius/gothic-1-community-patch/files/8009275/savegame3.zip)

I don't really care, I know there's nothing special behind that door. I am merely asking to know the status of that door - is it currently meant to be openable with this patch or not.

**Comment by AmProsius** [comment]
Thanks for reporting this issue. We will reopen this for review, because the door should be fixed already.

**Comment by szapp** [comment]
If I am not mistaken, the implementation doesn’t match the flow diagram in the issue description. Once saving and loading once acquired the key, the door’s key instance is no longer adjusted.

https://github.com/AmProsius/gothic-1-community-patch/blob/66e2fc2c1b10525ec27416debc509e5ceb409ff0/src/Ninja/G1CP/Content/Fixes/Gamesave/fix046_SmithDoor.d#L79-L82

**Comment by AmProsius** [comment]
Isn't it simply

```d
    // Check if new key item already exists somewhere
    if (!G1CP_IsItemInstantiated("G1CP_046_SmithDoor_Item")) {
        return FALSE;
    };
```

**Comment by szapp** [comment]
No quite the opposite actually. The code below that if-block creates the key if not present. What is incorrect in the script is that the key instance of the door is not adjusted and the function returns false, if the new key already exists.

**Comment by szapp** [comment]
@AmProsius The issue should be addressed in branch `bug046`. Would you be so kind and pull it to try:
1. Run the test once.
2. Pick up the key.
3. Save the game. *
4. Few variations:
    1. Do nothing and continue directly to step 5
    2. Load the just created gamesave
    3. Quit the game, restart it and load the gamesave
5. Run the test again.
6. Follow the on screen instructions of the test.

* In the zSpy you should find the fix being re-applied after saving is finished. This was not the case before the recent fix.

**Comment by AmProsius** [comment]
Confirmed. I will then close this issue.
